File: src/main/java/macro/buddy/gems/GemInfo.java
package macro.buddy.gems;

import java.util.SortedSet;

/**
 * Created by Macro303 on 2019-Nov-29.
 */
public class GemInfo implements Comparable<GemInfo> {
	private String name;
	private String slot;
	private SortedSet<GemTag> tags;
	private boolean hasVaal;
	private boolean hasAwakened;

	public GemInfo() {
	}

	public GemInfo(String name, String slot, SortedSet<GemTag> tags, boolean hasVaal, boolean hasAwakened) {
		this.name = name;
		this.slot = slot;
		this.tags = tags;
		this.hasVaal = hasVaal;
		this.hasAwakened = hasAwakened;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getSlot() {
		return slot;
	}

	public void setSlot(String slot) {
		this.slot = slot;
	}

	public SortedSet<GemTag> getTags() {
		return tags;
	}

	public void setTags(SortedSet<GemTag> tags) {
		this.tags = tags;
	}

	public boolean hasVaal() {
		return hasVaal;
	}

	public void setHasVaal(boolean hasVaal) {
		this.hasVaal = hasVaal;
	}

	public boolean hasAwakened() {
		return hasAwakened;
	}

	public void setHasAwakened(boolean hasAwakened) {
		this.hasAwakened = hasAwakened;
	}

	public String getFilename(boolean isVaal, boolean isAwakened) {
		String filename = getName().replaceAll(" ", "_");
		if (isVaal && hasVaal)
			filename += "[Vaal]";
		if (isAwakened && hasAwakened)
			filename += "[Awakened]";
		return filename + ".png";
	}

	@Override
	public String toString() {
		return "GemInfo{" +
				"name='" + name + '\'' +
				", slot='" + slot + '\'' +
				", tags=" + tags +
				", hasVaal=" + hasVaal +
				", hasAwakened=" + hasAwakened +
				'}';
	}

	@Override
	public int compareTo(GemInfo other) {
		return getName().compareToIgnoreCase(other.getName());
	}
}

File: src/main/kotlin/github/macro/build_info/gems/GemInfo.kt
package github.macro.build_info.gems

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleListProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.collections.FXCollections
import tornadofx.*

/**
 * Created by Macro303 on 2020-Jan-13.
 */
@JsonDeserialize(using = GemDeserializer::class)
class GemInfo(
	name: String,
	slot: String,
	tags: List<GemTag>,
	hasVaal: Boolean,
	hasAwakened: Boolean,
	acquisition: Acquisition
) {
	val nameProperty = SimpleStringProperty()
	var name by nameProperty

	val slotProperty = SimpleStringProperty()
	var slot by slotProperty

	val tagsProperty = SimpleListProperty<GemTag>()
	var tags by tagsProperty

	val hasVaalProperty = SimpleBooleanProperty()
	var hasVaal by hasVaalProperty

	val hasAwakenedProperty = SimpleBooleanProperty()
	var hasAwakened by hasAwakenedProperty

	val acquisitionProperty = SimpleObjectProperty<Acquisition>()
	var acquisition by acquisitionProperty

	init {
		this.name = name
		this.slot = slot
		this.tags = FXCollections.observableList(tags)
		this.hasVaal = hasVaal
		this.hasAwakened = hasAwakened
		this.acquisition = acquisition
	}

	fun getFilename(isVaal: Boolean, isAwakened: Boolean): String {
		var output = name.replace(" ", "_")
		if (isVaal && hasVaal)
			output += "[Vaal]"
		if (isAwakened && hasAwakened)
			output += "[Awakened]"
		return "$output.png"
	}

	override fun toString(): String {
		return "GemInfo(name=$name, slot=$slot, tags=$tags, hasVaal=$hasVaal, hasAwakened=$hasAwakened, acquisition=$acquisition)"
	}
}
